### PR TITLE
fix(chain): pin memtrie root across optimistic-apply queue

### PR DESCRIPTION
**Pull request by darioush**
An optimistic-block apply task can sit in the apply_chunks queue long enough that `update_flat_storage_and_memtrie` on the chain thread evicts the prev-state root from memtrie. When the queued apply runs it then can't find its root and the apply fails.

Fix: introduce `MemTrieRootPin` (RAII refcount holder on a memtrie root) and opaque wrapper `MaybePinnedMemtrieRoot`. The chain thread acquires the pin synchronously in `get_update_shard_job` (and equivalents in `chain_update`, `chunk_executor_actor`) before scheduling the apply task, captures it in `ApplyChunkShardContext`, and the runtime asserts the pin matches the apply's `(shard_uid, state_root)` at
`apply_chunk` entry. The pin's Drop releases the refcount when the apply task completes.

Includes a regression test `test_optimistic_apply_memtrie_gc_race` that uses a per-account artificial-delay knob on the apply_chunks spawner to reproduce the race deterministically in test-loop.

### CODE REVIEW

File: chain/chain/src/runtime/mod.rs [hunk 1]
```diff
@@ -1161,6 +1163,11 @@ impl RuntimeAdapter for NightshadeRuntime {
             .with_label_values(&[&apply_reason.to_string(), &shard_id.to_string()])
             .start_timer();
 
+        if matches!(storage_config.source, StorageDataSource::Db) {
```
Comment (wacban) [inline chain/chain/src/runtime/mod.rs]: nit: please avoid non-exhaustive matches - in case a new variant is added, it's better to be forced to handle it explicitly. If you don't want a full match statement here you can make it a method on the enum itself.

File: chain/chain/src/runtime/mod.rs [hunk 2]
```diff
@@ -1161,6 +1163,11 @@ impl RuntimeAdapter for NightshadeRuntime {
...
+            let shard_uid = self.get_shard_uid_from_prev_hash(shard_id, &block.prev_block_hash)?;
+            chunk.memtrie_pin.assert_pinned(&self.tries, shard_uid, &storage_config.state_root);
```
Comment (wacban) [inline chain/chain/src/runtime/mod.rs]: Are we guaranteed here that this is a new chunk? In case of a missing chunk at the resharding boundary you would get shard_uid to be that of the parent which most likely won't be pinned.

Comment (wacban) [reply chain/chain/src/runtime/mod.rs]: I'll ask claude to check other usages of `get_shard_uid_from_prev_hash` just in case.

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -1775,6 +1775,11 @@ impl Chain {
     ) {
         // Track all children using `parent_span`, as they may be processed in parallel.
         let parent_span = Span::current();
+        // Distinguish apply tasks by block kind so tests can delay one variant.
+        let task_name = match &block {
+            BlockToApply::Optimistic(_) => "apply_chunks_optimistic",
+            BlockToApply::Normal(_) => "apply_chunks_normal",
+        };
```
Comment (wacban) [inline chain/chain/src/chain.rs:1790]: This is lovely!

File: core/store/src/trie/shard_tries.rs [hunk 4]
```diff
@@ -620,6 +620,24 @@ impl ShardTries {
...
+    /// no memtrie is loaded or `state_root` is empty; errors if memtrie is
+    /// loaded but the root is missing.
+    pub fn maybe_pin_memtrie_root(
+        &self,
+        shard_uid: ShardUId,
+        state_root: StateRoot,
+    ) -> Result<MaybePinnedMemtrieRoot, StorageError> {
+        if state_root == CryptoHash::default() {
+            return Ok(MaybePinnedMemtrieRoot::no_memtries());
+        }
```
Comment (wacban) [inline core/store/src/trie/shard_tries.rs:636]: Why do we need this condition? When does it happen?

File: test-loop-tests/src/setup/setup.rs [hunk 5]
```diff
@@ -160,8 +161,15 @@ pub fn setup_client(
     // Make sure this is the same as the account_id of the client to redirect the network messages properly.
     let peer_id = PeerId::new(create_test_signer(account_id.as_str()).public_key());
 
+    // Default per-task virtual delay for the test-loop spawner. Tests can
+    // override per `(account, task_name)` via `TestLoopBuilder::task_delay_fn`.
+    const DEFAULT_TASK_DELAY: Duration = Duration::milliseconds(80);
```
Comment (wacban) [inline test-loop-tests/src/setup/setup.rs:166]: Just for my education, this is a virtual delay, not a wall clock delay right?

File: core/store/src/trie/mem/memtries.rs [hunk 6]
```diff
@@ -184,7 +186,22 @@ impl MemTries {
...
+    /// Pair every successful call with exactly one [`Self::delete_root`].
+    pub(crate) fn add_root_ref(&mut self, state_root: &CryptoHash) -> Result<(), StorageError> {
+        let ids = self.roots.get(state_root).ok_or_else(|| {
+            StorageError::StorageInconsistentState(format!(
+                "failed to find root node {:?} in memtrie when adding ref",
+                state_root
+            ))
+        })?;
+        let last_id = *ids.last().unwrap();
+        last_id.add_ref(self.arena.memory_mut());
```
Comment (wacban) [inline core/store/src/trie/mem/memtries.rs:196]: The invariant from [this](https://github.com/near/nearcore/blob/e25af50454f81dffbfb07af3d094fca9bcc294f1/core/store/src/trie/mem/memtries.rs#L33-L38) comment is no longer true. I'll leave it to you to figure out if this needs fixing or just updating the comment :)

**Review by wacban — COMMENTED** [review]
LGTM but I'll let others have a look and approve. Few comments from my side.

File: test-loop-tests/src/setup/setup.rs (hunk 5 above)
Comment (darioush) [reply test-loop-tests/src/setup/setup.rs:166]: That's correct, it's a test-loop style virtual delay.

File: core/store/src/trie/mem/memtries.rs (hunk 6 above)
Comment (darioush) [reply core/store/src/trie/mem/memtries.rs:196]: MemTries has three kinds of "refcount holders" for a state root R:
  - entries in heights (one push per insert_root call)
  - snapshot_root if it equals R
  - live MemTrieRootPins (new in this PR)

Each holder calls add_ref once when it starts, and calls delete_root(R) once when it ends. So at any moment:
```
  sum(refcount(node) for node in roots[R]) == (# heights occurrences of R) + (1 if snapshot==R) + (# live
  pins on R)
```

Updated comment in https://github.com/near/nearcore/pull/15695/commits/5094185ba9f4f5b934d17b8b47d9dd2d4a5f92a0

File: chain/chain/src/runtime/mod.rs (hunk 2 above)
Comment (darioush) [reply chain/chain/src/runtime/mod.rs]: Fixed by plumbing shard_uid to ApplyChunkShardContext in https://github.com/near/nearcore/pull/15695/commits/e8c23e98ad62f7771a38fe5ee32bdba053bf99a6, PTAL

File: core/store/src/trie/shard_tries.rs (hunk 4 above)
Comment (darioush) [reply core/store/src/trie/shard_tries.rs:636]: Seems when the trie is empty, we don't have any node corresponding to it in memtries (see `insert_root`).
This can happen in some edge cases where the shard has an empty state, like a missing chunk after genesis.
I observed it in some tests like `test_load_memtrie_after_empty_chunks`

File: chain/chain/src/chain.rs [hunk 7]
```diff
@@ -3398,6 +3404,13 @@ impl Chain {
         tracing::debug!(target: "chain", %shard_id, ?cached_shard_update_key, "creating shard update job");
 
         let mut on_post_state_ready = None;
+        // Pin the prev-state root so memtrie GC can't free it while the apply
+        // task sits in the queue.
+        let prev_state_root = if is_new_chunk {
+            chunk_header.prev_state_root()
+        } else {
+            *self.get_chunk_extra(prev_hash, &shard_context.shard_uid)?.state_root()
```
Comment (VanBarbascu) [inline chain/chain/src/chain.rs]: Worth mentioning as the logic is on the hot path. Duplicate `get_chunk_extra` for `is_new_chunk = false`: the same `self.get_chunk_extra(prev_hash, &shard_context.shard_uid)?` call already happens inside the `else` arm of `shard_update_reason` just below (where it's bound to `prev_chunk_extra` for `OldChunkData`). Worth folding so we read once and reuse — e.g. compute `prev_chunk_extra` early when `!is_new_chunk`, take `*prev_chunk_extra.state_root()` for the pin, and move the same value into the `OldChunkData`.

File: test-loop-tests/src/tests/optimistic_block.rs [hunk 8]
```diff
@@ -341,3 +342,45 @@ fn test_optimistic_block_with_invalidated_outcome() {
...
+#[cfg_attr(feature = "protocol_feature_spice", ignore)]
+#[test]
+fn test_optimistic_apply_memtrie_gc_race() {
+    let num_shards = 3;
+
+    // Delay one node's `apply_chunks_optimistic` by several block intervals
+    // so `last_final_block` advances past the optimistic block's prev-height
+    // while the apply task is still queued.
+    let slow_node: AccountId = "account0".parse().unwrap();
+    let metric_before = chain_metrics::NUM_FAILED_OPTIMISTIC_BLOCK_APPLIES.get();
```
Comment (VanBarbascu) [inline test-loop-tests/src/tests/optimistic_block.rs]: Flake risk. ([ref](https://doc.rust-lang.org/book/ch11-02-running-tests.html#running-tests-in-parallel-or-consecutively)) `NUM_FAILED_OPTIMISTIC_BLOCK_APPLIES` is a process global `LazyLock<IntCounter>`. Cargo runs tests in the same binary in parallel, so any concurrent test in this crate that triggers an optimistic-apply failure will fail this assertion. 

Running this test in serial mode is not a great option, but if no alternative is straight forward, we can take the cost. 

At minimum, please confirm no other test in `test-loop-tests` currently exercises the optimistic-apply failure path.

File: core/store/src/trie/mem/memtries.rs [hunk 9]
```diff
@@ -257,6 +275,64 @@ impl MemTries {
...
+}
+
+/// Opaque handle for the runtime; validated at the entry of `apply_chunk`.
+pub struct MaybePinnedMemtrieRoot {
+    pin: Option<MemTrieRootPin>,
+}
+
+impl MaybePinnedMemtrieRoot {
+    /// For callers whose apply path doesn't go through memtrie. Panics in
+    /// [`Self::assert_pinned`] if memtrie is actually loaded for the shard.
```
Comment (VanBarbascu) [inline core/store/src/trie/mem/memtries.rs]: Memtries can be loaded but sometimes you want to use the recorded storage.
```suggestion
    /// [`Self::assert_pinned`] if memtrie is actually used for the shard processing.
```

File: core/store/src/trie/mem/memtries.rs [hunk 10]
```diff
@@ -257,6 +275,64 @@ impl MemTries {
...
+        // Empty trie has no memtrie entry (see `MemTries::insert_root`),
+        // so the matching pin in `ShardTries::maybe_pin_memtrie_root`
+        // short-circuits to `no_memtries()` and there's nothing to check.
+        if state_root == &CryptoHash::default() {
+            return;
+        }
+        let Some(pin) = &self.pin else {
+            assert!(
+                tries.get_memtries(shard_uid).is_none(),
+                "apply_chunk for {shard_uid:?}: memtrie is loaded but no pin held; \
```
Comment (VanBarbascu) [inline core/store/src/trie/mem/memtries.rs]: Worth adding the `state_root` to help link it to the chunk that is in processin.

File: core/store/src/trie/mem/memtries.rs [hunk 11]
```diff
@@ -257,6 +275,64 @@ impl MemTries {
...
+                 call ShardTries::maybe_pin_memtrie_root() instead of no_memtries()",
+            );
+            return;
+        };
+        assert_eq!(pin.shard_uid, shard_uid, "memtrie pin shard_uid mismatch");
+        assert_eq!(&pin.state_root, state_root, "memtrie pin state_root mismatch");
```
Comment (VanBarbascu) [inline core/store/src/trie/mem/memtries.rs]: Does this check need to be so strict? Is it worth crashing over this? No strong opinion, just want to understand.

**Review by VanBarbascu — COMMENTED** [review]
Couple of questions and suggestions.

File: core/store/src/trie/mem/memtries.rs (hunk 11 above)
Comment (darioush) [reply core/store/src/trie/mem/memtries.rs]: Let's switch to debug_assert + error log for now [7dfeebe](https://github.com/near/nearcore/pull/15695/commits/7dfeebe5e6a0a723a88ab2c4ff5335d6ce019146)

File: test-loop-tests/src/tests/optimistic_block.rs (hunk 8 above)
Comment (darioush) [reply test-loop-tests/src/tests/optimistic_block.rs]: Good point, added a counter for it gated by test_features. [32f54ec](https://github.com/near/nearcore/pull/15695/commits/32f54ece8036f6a66b3f7c991306f1ed7f27cbe4)

File: chain/chain/src/chain.rs (hunk 7 above)
Comment (darioush) [reply chain/chain/src/chain.rs]: Moved some code around to avoid this duplication in [3b25e0c](https://github.com/near/nearcore/pull/15695/commits/3b25e0cf87df547987d0156892a99c6a13817209)

**Review by VanBarbascu — APPROVED** [review]
LGTM!

I would appreciate if @Wiezzel takes a look at the impact around the rehsarding boundary. Main points: how do we manage the tries during reshardings? Do we need to worry about pinning both parent and descendant shards at the resharding boundary?

**Review by ssavenko-near — APPROVED** [review]

File: core/store/src/trie/mem/memtries.rs [hunk 12]
```diff
@@ -257,6 +275,83 @@ impl MemTries {
...
+            debug_assert_eq!(&pin.state_root, state_root, "memtrie pin state_root mismatch");
+        }
+    }
+}
+
+impl Drop for MemTrieRootPin {
+    fn drop(&mut self) {
+        self.memtries.write().delete_root(&self.state_root);
+    }
+}
```
Comment (Wiezzel) [inline core/store/src/trie/mem/memtries.rs:337]: Is it possible for `MemTrieRootPin` to survive over memtrie resharding? If yes, I believe that could call `delete_root` with a non-existent root (because the parent memtrie has been already removed).

Comment (darioush) [reply core/store/src/trie/mem/memtries.rs:337]: Indeed this is the case, with some efforts I could reproduce it in a test-loop https://github.com/darioush/nearcore/pull/15/changes.

I won't commit that test though, I'll keep a unit test and patch the behavior to check for the root before deleting with a warning log if the root is not found. This should be enough to notice it if it keeps happening but not hit `debug_assert!` in the `delete_root` path: https://github.com/near/nearcore/pull/15695/changes/1e24598517d836d327cb360359a1ea6dfaa4d5c1

Comment (darioush) [reply core/store/src/trie/mem/memtries.rs:337]: @Wiezzel unfortunately, I don't think just patching the `delete_root` case is sufficient. As the apply_chunk may access the Trie it creates at the [beginning](https://github.com/near/nearcore/blob/9e5054d434461ef00a1e66af1c7fbd57816244bc/chain/chain/src/runtime/mod.rs#L1170-L1182).

Do you think [this](https://github.com/near/nearcore/pull/15738/changes#diff-ba2b649b7a9490c5c6cc1adb9c8b34da9c92bbde788df9a6b9635d545f57235fR682-R683) is a valid way to allow in-flight applications to complete over the memtrie resharding boundary?

Comment (Wiezzel) [reply core/store/src/trie/mem/memtries.rs:337]: @darioush Yes, I think this should do the trick. Approved #15738

Why not commit `slow_test_resharding_v3_pin_survives_freeze` though?

Comment (darioush) [reply core/store/src/trie/mem/memtries.rs:337]: I think the unit test in #15378 shows the same issue without extending the `resharding_v3` framework, but if you like the test I am happy to add it
